src/Control/Algebra.hs

```diff
@@ -167,43 +173,49 @@ toRWSTF w (a, s, w') = RWSTF (a, s, mappend w w')
 {-# INLINE toRWSTF #-}
 
 instance (Algebra sig m, Effect sig, Monoid w) => Algebra (Reader r :+: Writer w :+: State s :+: sig) (RWS.Lazy.RWST r w s m) where
-  alg (L (Ask       k))      = RWS.Lazy.ask >>= k
-  alg (L (Local f m k))      = RWS.Lazy.local f m >>= k
-  alg (R (L (Tell w k)))     = RWS.Lazy.tell w *> k
-  alg (R (L (Listen m k)))   = RWS.Lazy.listen m >>= uncurry (flip k)
-  alg (R (L (Censor f m k))) = RWS.Lazy.censor f m >>= k
-  alg (R (R (L (Get   k))))  = RWS.Lazy.get >>= k
-  alg (R (R (L (Put s k))))  = RWS.Lazy.put s *> k
-  alg (R (R (R other)))      = RWS.Lazy.RWST $ \ r s -> unRWSTF <$> alg (thread (RWSTF ((), s, mempty)) (\ (RWSTF (x, s, w)) -> toRWSTF w <$> RWS.Lazy.runRWST x r s) other)
+  alg hom = \case
+    L (Ask       k)      -> RWS.Lazy.ask >>= hom . k
+    L (Local f m k)      -> RWS.Lazy.local f (hom m) >>= hom . k
+    R (L (Tell w k))     -> RWS.Lazy.tell w *> hom k
+    R (L (Listen m k))   -> RWS.Lazy.listen (hom m) >>= hom . uncurry (flip k)
+    R (L (Censor f m k)) -> RWS.Lazy.censor f (hom m) >>= hom . k
+    R (R (L (Get   k)))  -> RWS.Lazy.get >>= hom . k
+    R (R (L (Put s k)))  -> RWS.Lazy.put s *> hom k
+    R (R (R other))      -> RWS.Lazy.RWST $ \ r s -> unRWSTF <$> alg id (thread (RWSTF ((), s, mempty)) (\ (RWSTF (x, s, w)) -> toRWSTF w <$> RWS.Lazy.runRWST (hom x) r s) other)
```

Comment:
I also took the opportunity to rewrite _all_ the algebras using `LambdaCase`, because

1. I find `case` easier to read than all the chaff on the left, and

2. `alg hom x = case x of …` is the quintessential case for `\case`.